Add version and spec release information to release notes

olavphibm opened the pull request.

Commit:
Create class to hold spec file specific functions

Commit:
Use Python library to compare RPM versions

This avoids running an external command.

Commit:
Add version and release properties to RPM_Package

This makes it easier to query those specific attributes from the spec
file.

Remove those same properties from Version class, they were not being
used.

Commit:
Add version and release information to release notes

Commit:
Add options to set release notes repository URL and branch

This allows the checkout repository to be different than the one where
the created release notes are pushed.

farosas commented on lib/rpm_package.py:
```diff
@@ -130,3 +130,11 @@ def _load(self):
+
+    @property
+    def version(self):
+        return self.spec_file.query_tag("version")
```
We will start a process and open the file every time we check this property. I think we should cache this value.

olavphibm replied:
Agreed. Added a cache to the `query_tag` method.

farosas commented on lib/rpm_package.py:
```diff
@@ -130,3 +130,11 @@ def _load(self):
+
+    @property
+    def release(self):
+        return self.spec_file.query_tag("release").split('.')[0]
```
Same here.

Also, the release field should not be meaningful in the context of an rpm without the whole string. That property was not being used in upgrade_versions.py so either remove it or return the whole Release string.

olavphibm replied:
Ok, returning the whole string since it will be used in create_release_notes.py.

farosas commented on lib/rpm_package.py:
```diff
@@ -15,16 +15,81 @@
+        LOG.info("Updating '%s' version to: %s" % (self.path, new_version))
+
+        with open(self.path, 'r+') as f:
+            content = f.read()
```
Opening the file here is not optimal. I'm adding code to update the 'prerelease' tag and this will be duplicated. Perhaps we should have `read_spec` and `write_spec` methods and keep the content of the file in an attribute so that we read it and write it only once. We could also delay the reading until a functions needs it. What do you think?

farosas replied:
fyi: https://github.com/fabianorosas/builds/commits/prerelease

olavphibm replied:
I've created a property `content` to handle content manipulation and a method `write_content` to commit to the file.

farosas commented on upgrade_versions.py:
```diff
@@ -130,9 +104,10 @@ def update(self, user_name, user_email):
+            self._spec_version, self._repo_version)
         if result < 0:
-            self._update_version()
+            pkg.spec_file.update_version()
```
Missing the version here.

farosas commented on upgrade_versions.py:
```diff
@@ -186,13 +129,12 @@ def _bump_release(self, pkg, log=None, user_name=None, user_email=None):
-            rpm_bump_spec(pkg.specfile, log, user_name, user_email)
+            pkg.spec_file.bump_release(log, user_name, user_email)
 
     def _read_spec(self):
```
I don't think this method makes sense anymore. Just make the calls directly to `self.pkg.version`/`release` or move these attributions to the constructor.

olavphibm replied:
Agreed. Changed to `self.pkg.version/release`

farosas reviewed (changes requested):
Good job creating the SpecFile class, it's a nice improvement over the previous code.

Commit:
Create methods for getting, setting and writing spec file content

This allows more control on when to actually read/write the file when
it is modified by multiple methods.

Commit:
Cache tag queries made to the spec file

farosas commented on lib/rpm_package.py:
```diff
@@ -15,16 +16,103 @@
+        """
+        self._content = value
+
+    def write_content(self):
```
Shouldn't this method reset `self._content` as well?

farosas replied:
Ok, but we might then introduce a bug when using rpmdev-bumpspec which also writes to the spec file. Calling `bump_release` in between calls to `write_content` would lose the bump information.

olavphibm replied:
Good catch. I've forced the bump_release method to set `self._content = None`, so it gets reloaded when needed.

farosas commented on lib/rpm_package.py:
```diff
@@ -15,16 +16,117 @@
+
+    def _replace_macro_definition(self, macro_name, replacement, text):
+        return re.sub(r'(%%define\s+%s\s+)\S+' % macro_name,
+                      r'\g<1>' + replacement, text)
```
We could remove the `text` argument and use `self.content` directly here.
